pkg/ksym/optimizer.go

```diff
@@ -0,0 +1,292 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ksym
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"unsafe"
+
+	"golang.org/x/exp/mmap"
+)
+
+// There are several use-cases, such as symbolization, that conceptually boil
+// down to a list of tuples, each formed of an identifier and a string, where
+// we want to efficiently find the string for a particular identifier.
+//
+// A possible approach to solve this problem is to create a large list
+// in memory, sorted by identifier, so we can binary-search over the ids
+// and find the entry for which entry_i.Id <= Id < entry_i+1.Id.
+//
+// While this works well and it's easy to understand and maintain, when
+// the list grows too large, this can be a large source of retained memory.
+// This issue is particularly bad when memory offloading (swap/zswap) is not
+// enabled, as the "cold" anonymous memory won't have a way to be moved to
+// secondary storage.
+//
+// This implementation produces a simple binary format that's easy to write
+// and read, but most importantly, it should be efficient to query.
+//
+// ┌─────────┬────────────────────────────┬────────────────────────────────────────────┐
+// │         │                            │                                            │
+// │ Header  │  Strings with nul endings  │  Sorted ids + meta information on strings  │
+// │         │                            │                                            │
+// └─────────┴────────────────────────────┴────────────────────────────────────────────┘
+//
+// The strings aren't deduplicated or optimized in any way, to reduce the memory
+// usage during the write phase.
+//
+// The file is read with `mmap(2)`, to avoid performing any read system calls
+// while binary searching over the identifiers, and leveraging the caching layer
+// of the filesystem. As we now have a backing file, rather than being anonymous
+// memory, the OS can remove cached pages if there's need for more memory.
+
+const (
+	MAGIC   = uint32(0x8A4CA)
+	VERSION = uint32(1)
+)
+
+var (
+	errSymbolNotFound   = errors.New("symbol not found")
+	errAlreadyFinalized = errors.New("already finalized")
+	errBadMagic         = errors.New("bad magic identifier")
+	errBadVersion       = errors.New("bad version")
+	errReadZeroBytes    = errors.New("read zero bytes")
+)
+
+type fileHeader struct {
+	Magic           uint32
+	Version         uint32
+	AddressesOffset uint32
+	AddressesCount  uint32
+}
+
+type entry struct {
+	address uint64
+	offset  uint32
+	len     uint32
```

Comment:
if we null-terminate the strings, why not just read until we hit the null? we'd save 25% of size of entries

Reply:
I've been thinking about this and I like having the size so we can pre-allocate the buffer to read the bytes that make the string with the right size.

But now that you mention this, I was thinking that perhaps we can use a `uint8` for the len, what do you think?

Reply:
What I've done elsewhere is write the largest string ever seen into the header, and allocate that size. I'm pretty sure the worst thing we might allocate with kernel strings is 1k bytes (and probably much less).

Reply:
Got it! I agree that the maximum length kernel symbol won't be too large. I think I will keep the code like this because I would like to make the read path allocate (and copy) as little memory as possible. But I am happy to continue exploring optimizations if the kernel symbolizer is not as efficient as we expect in production! 😄 

Meanwhile, I have changed the `len` to use 2 bytes instead of 4 bytes, while I hope there won't be kernel symbols of more than 2**8 = 256 characters, I wanna be sure we don't hit that. Some compilers / post compiler optimizers change function names and add suffixes, increasing some function's names

Reply:
sgtm